src/libsyntax_ext/assert.rs

```diff
@@ -23,7 +23,9 @@ pub fn expand_assert<'cx>(
         }
     };
 
-    let sp = cx.with_legacy_ctxt(sp);
+    // `core::panic` and `std::panic` are different macros, so we use call-site
+    // context to pick up whichever is currently in scope.
+    let sp = cx.with_call_site_ctxt(sp);
```

Comment:
I'd rather use the call-site span in a more targeted way though, for the `panic` identifiers specifically, and def-site for everything else.
(Not sure if this makes observable difference in this case, but it's a safer default.)